### DISCUSSION TITLE
Auth - Which auth service is being used behind the Microsoft Auth Provider?

### DISCUSSION THREAD

**Discussion by JeremyQtweb** [supabase/supabase#28755] [Questions] [answered]
Hey team,

We've been testing the new Microsoft Entra External ID (EEID) service, their replacement for Azure AD B2C.

Presumably the Supabase Auth - Microsoft Provider is one of those services.

Can I confirm with someone if that is true, and which one the Microsoft Provider is using?

**Comment by ssudaraka** [comment]
Hi @JeremyQtweb,

Microsoft Entra is just a new rebranding of Azure AD. The features and the infrastructure around the product are the same. You can use Entra as a Supabase auth provider.

**Reply by JeremyQtweb** [reply]
Uhhh, it's not actually the same, but I need to be more explicit.

Microsoft Entra ID is effectively a rebranding of Azure AD, correct.

What I am actually asking about though, is the new Customer Identity and Access Management service called [Microsoft External Entra ID for Customers](https://learn.microsoft.com/en-us/entra/external-id/customers/).

While [this answer](https://learn.microsoft.com/en-us/answers/questions/1556632/confusion-around-azure-ad-b2c-vs-microsoft-entra-e) is lacking in technical differences, from our testing and conversations with Microsoft, we know there are technical differences from the AADB2C service, and they are proving to be incompatible with Supabase's integration accordingly.

Effectively, External Entra ID is not feature complete nor fit-for-purpose yet. So this is not technically a Supabase issue, but if users are trying to auth through Supabase using a EEID tenant, but the MS provider Supabase talks to is the AADB2C service, pretty sure it ain't gonna work correctly.

For one, MFA does not work with all account types, and only supports Email OTP currently. The way that Microsoft handles accounts in External Entra ID has changed and thus does not behave like AADB2C does.

One of the most obvious clues that something has changed is in the name, the acronym CIAM. While Supabase is currently configured with a Azure Tenant URL of `https://login.microsoftonline.com/{tenant-id}` (AADB2C), Microsoft EEID requires the use of `https://{tenant-url}.ciamlogin.com`.

Further, Microsoft provides a set of sample applications for EEID of which we've tested the [React app](https://learn.microsoft.com/en-nz/entra/external-id/customers/sample-single-page-app-react-sign-in) version. In testing, using said application via their MSAL library we managed to get our external customers to login and use Email OTP for MFA via EEID, but our federated "workforce" accounts (internal) were unable to login at all.

**So in summary, these are the issues we face:**

- Supabase's MS auth provider configured with login.microsoft.com needs accounts to be “workforce” accounts to work at all (can login, current implementation), but we cannot utilise any MFA (internal & external users).
- Supabase's MS auth provider configured with caimlogin.com works with non “workforce” accounts (external), which is what we want, but has issues dealing with MFA and federated "workforce" accounts (internal) cannot login at all.
- A plain old React app with stock standard MSAL library and configured with ciamlogin.com works with non “workforce” accounts (external) and allows MFA, but federated "workforce" accounts (internal) cannot login at all).

We believe one change that might resolve the issues - at least in part - is having a second Microsoft Provider in Supabase which runs against the Microsoft EEID for Customers service, rather than AADB2C as presumably the current provider does.

I can provide the error messaging we receive from MS when the various failure scenarios above occur, but be warned, some of the error messages/codes are still not recognised by Microsoft's own error messaging documentation. Sigh.

I hope this additional info clarifies the issues we're facing.

Cheers.

**Reply by ssudaraka** [reply]
@JeremyQtweb Extremely sorry, I might have overlooked the information available on their documentation. Let me investigate this more and get back to you soon!

**Reply by JeremyQtweb** [reply]
No worries at all, when is MS documentation/product identity not confusing!

**Reply by JeremyQtweb** [reply]
Hi @ssudaraka ,

Unfortunately for us this is now a Production issue, so I'm following up to see if you've had a chance to confirm my suspicion and determine if there are any plans to address it?

Thanks for your assistance.

**Reply by ssudaraka** [reply]
Hi @JeremyQtweb,

I apologize for the extremely delayed response; I was away most of the last month. 

I tried to sign up for a Microsoft Entra ID P2 tenant to test this behavior, but the endpoints it's giving me for integrating auth has the microsoftonline.com domain. Can you tell me more about how to obtain ciamlogin.com URLs for Entra ID applications?

Also, you can use SAML with your Entra ID tenant if it doesn't work with OAuth. Maybe that will help?

**Reply by JeremyQtweb** [reply]
Hi @ssudaraka ,

Entra ID is a "internal use only product".
 
Azure AD B2C is an "external product" which has recently been superseded by **Entra External ID**. AADB2C has been sunset, will remain in production for ~5 years they say, but its not "the future" so it will receive no more development or support from MS - and "most new installs of external MS directories" will be **Entra External ID**.
 
**Entra External ID** for Customers (an external product) works a bit like AADB2C with the key difference being the URLs and the claim/access tokens used.

**Reply by JeremyQtweb** [reply]
Bump

**Reply by akarray** [reply]
Any plan ? 
EEID is a game changer !

**Comment by hf** [comment]
These are the issuers supported currently. https://github.com/supabase/auth/blob/master/internal/api/provider/azure.go#L16-L42

Please link to documentation from Microsoft about this new system they have, as a quick search online doesn't yield any meaningful tech documentation. Just marketing word salad.

Contributions to the Auth repo are welcome!

**Reply by JeremyQtweb** [reply]
Hi @hf ,

Thank you for responding.

Microsoft has a dedicated section for [External Entra ID documentation](https://learn.microsoft.com/en-us/entra/external-id/customers/).

[Code examples for EEID](https://developer.microsoft.com/en-us/identity/external-id) are available too.

Cheers,
Jeremy

**Comment by hf** [comment]
Can't you do this by specifying the CIAM URL here in Azure Tenant? That's what the code supports today.

<img width="868" alt="image" src="https://github.com/user-attachments/assets/160f9af7-0d44-4009-b323-253feaca0ef9" />

**Reply by JeremyQtweb** [reply]
Apologies for the delay @hf , have been on vacation and then sick. Sigh.

Trying the CIAM URL was one of the first things we tried.

Here is a complete summary from both Microsoft and our own team on where we are at right now, and what we've tried...


**What we are trying to do:**
- Sign in External and Internal (federated) users (with MFA of some form) via:
	- Microsoft Entra External ID (EEID) CIAM tenant
	- Supabase Azure Provider
	- MFA via email OTP for External users

**Outcome:**
- Using https://login.microsoftonline.com as Tenant URL 
	- Users of either type can sign in
	- MFA does not work
- Using https://{tenant-url}.ciamlogin.com as Tenant URL
	- Users of either type _**cannot**_ sign in



### Summary of the case, from Microsoft’s point of view

As soon as a CONSUMER user i.e. a user created via a user flow or a user created via Graph API calls; 

User is added to a role. We stop treating the user as CONSUMER as the user now has some extra permissions (for example, Directory readers role giving some basic information of other users) thus we treat the user as a member of ENTERPRISE.

We fallback to legacy workforce-based scenarios and one of them is that they can't use Email OTP as MFA and also, they have permission to use Microsoft first party apps like my*, which allows them to register more authentication methods. 

On a CIAM tenant, there are 2 types of users, CONSUMER users which has no roles, and they will get consumer experience and get token only from CIAM apps on CIAM host endpoint i.e. ciamlogin.com. Other is more of admin persona with some role assigned, they will be able to access Microsoft first party app and they will see workforce experience like Email OTP not allowed as MFA, access to portal and way to register new MFA methods from my*

These are CONSUMER users which lack permission to register new MFA Methods and fail that's why we are seeing the failure. Users with directory roles don't have permission to do Email OTP because they are now treated as organizational accounts.

So we need to separate CONSUMER users and ENTERPRISE that might be invited users to give access to some applications.

In this case, B2C/CIAM should be using either b2clogin.com/ciamlogin.com or a custom domain.
When the user logs in via login.microsoftonline.com we expect the user to be having a Role assigned.
If a CIAM user with a role assigned attempts MFA, it will fail as explained above.

[Tenant configurations and External ID - Microsoft Entra External ID](https://learn.microsoft.com/en-us/entra/external-id/tenant-configurations)

[User default permissions in external tenants - Microsoft Entra External ID](https://learn.microsoft.com/en-us/entra/external-id/customers/reference-user-permissions)

 
**Additional:**

Currently, there hasn’t been an official timeline announced for adding Authenticator app support to CIAM tenants, as this feature is still under development by the production team. At present, Microsoft Entra External ID (CIAM) supports multifactor authentication primarily through email OTP. Once the Authenticator app support is released, the official documentation will be updated accordingly: [Add multifactor authentication (MFA) to a customer app - Microsoft Entra External ID](https://learn.microsoft.com/en-us/entra/external-id/customers/how-to-multifactor-authentication-customers)

Additionally, the team is also gathering feedback for product development. If you have any suggestions or feedback on Azure product features, we can always share them via this link:

https://feedback.azure.com/d365community/forum/22920db1-ad25-ec11-b6e6-000d3a4f0789  

 

### Summary of the case, from Our point of view.

The biggie on why we were having problems is that we discovered an initial error went away by adding the users to a Directory Role, as explained above. Adding the user to any role now makes the user “Workforce”, think Work/School Account, and therefore the only supported MFA method for EEID users no longer works. If we take this role away everything behaves as expected.

Now, this brings to our second issue. When we take away the role our application login does not work any more (through Supabase). Test logins using Microsoft’s test process are fine.

Now this is easily explained from another line above when it talks about using b2clogin.com/ciamlogin.com instead of login.microsoftonline.com.

The Supabase auth config is using login.microsoftonline.com, not ciamlogin.com. So it makes perfect sense our users need to be Workforce type accounts for the logon process to work.

When we change the supabase to be using ciamlogin.com, logins don’t work at all. We believe the Microsoft Azure auth module/library in Supabase is not compatible with EEID.

We also spun up a pretty basic test React app using Microsofts MSAL library (bypassing Supabase). Now we had different behaviour here, with a plain EEID local account (a gmail email address) we could actually login fine to the React test app. Perfect, one would think. However, we then had 2 further problems: when we enabled MFA for the EEID local user the previously working login process stops working, and when we used a federated account it didn’t login at all.

Now this React module/library which is using the standard MSAL library came from a sample specifically for EEID users. So it appears this library is maybe not new enough to cope with all the features yet (circa August 24).

**So we have these 3 scenarios:**
- Supabase MS Auth provider needs accounts to be “workforce” accounts to work at all, but then as Workforce accounts in EEID we cannot utilise any MFA
- Supabase MS auth provider configured with caimlogin.com works with non “workforce” accounts, which is what we want, but has issues dealing with MFA and federated accounts
- Plain old React app with stock standard current version of MSAL works fine with ciamlogin.com but not dealing with the federated accounts. This is just for a comparison.


_Now - we completely understand that EEID is new - or what marketing likes to call 'cutting edge' - so we were bound to encounter some issues. However, our underlying belief is that the current Microsoft Auth Provider in Supabase has been developed against the original Azure AD B2C tenant (login.microsoft.com) and is not fundamentally compatible with EEID, purely due to the way Microsoft handles users differently than B2C did._

_Our assumption is that, whatever the token is that you guys need to pass back n forth with Microsoft, is slightly different with EEID than it is with B2C leading to the unexpected behaviours we're experiencing using the B2C tenant URL, and the inability to sign in at all with the CIAM tenant URL._

**Reply by JeremyQtweb** [reply]
Bump @hf @ssudaraka - any updates on this folks?

**Reply by JeremyQtweb** [reply, marked as answer]
@hf - I found [your PR](https://github.com/supabase/auth/pull/1541) in April last year where you added support for people using the CIAM based login i.e. the very thing I am asking about in this question (External Entra ID).

So, it appears that you do already 'support' this, but it's not working correctly for one of our use cases.

After further analysis we believe the issue is not one for Supabase to resolve.

We don't think EEID is doing what Microsoft thought it would do for our Linked Accounts, effectively treating them as Workforce accounts, which then eliminates the user of MFA (at the time this issue was identified).

There are new Authentication methods available now, which may solve our key issue. Otherwise, we may need to adjust our solution to support two authentication workflows (Internal vs External users).

Thank you for your time, and all that you guys do to maintain these packages etc.
It is appreciated.
